pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,604 @@
+	}
+	if vec.GetType() != o.GetType() {
+		return false
+	}
+	if vec.Nullable() != o.Nullable() {
+		return false
+	}
+	if vec.HasNull() != o.HasNull() {
+		return false
+	}
+	if vec.HasNull() {
+		if !vec.NullMask().Equals(o.NullMask()) {
+			return false
+		}
+	}
+	mask := vec.NullMask()
+	for i := 0; i < vec.Length(); i++ {
+		if mask != nil && mask.ContainsInt(i) {
+			continue
+		}
+		var v T
+		if _, ok := any(v).([]byte); ok {
+			if !bytes.Equal(vec.Get(i).([]byte), o.Get(i).([]byte)) {
+				return false
+			}
+		} else if _, ok := any(v).(types.Decimal64); ok {
+			d := vec.Get(i).(types.Decimal64)
+			od := vec.Get(i).(types.Decimal64)
+			if d.Ne(od) {
+				return false
+			}
+		} else if _, ok := any(v).(types.Decimal128); ok {
+			d := vec.Get(i).(types.Decimal128)
+			od := vec.Get(i).(types.Decimal128)
+			if d.Ne(od) {
+				return false
+			}
+		} else if _, ok := any(v).(types.TS); ok {
+			d := vec.Get(i).(types.TS)
+			od := vec.Get(i).(types.TS)
+			if types.CompareTSTSAligned(d, od) != 0 {
+				return false
+			}
+		} else if _, ok := any(v).(types.Rowid); ok {
+			d := vec.Get(i).(types.Rowid)
+			od := vec.Get(i).(types.Rowid)
+			if types.CompareRowidRowidAligned(d, od) != 0 {
+				return false
+			}
+		} else {
+			if vec.Get(i) != o.Get(i) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (vec *CnTaeVector[T]) ForeachWindow(offset, length int, op ItOp, sels *roaring.Bitmap) (err error) {
+	err = vec.forEachWindowWithBias(offset, length, op, sels, 0)
+	return
+}
+
+func (vec *CnTaeVector[T]) forEachWindowWithBias(offset, length int, op ItOp, sels *roaring.Bitmap, bias int) (err error) {
+	if sels == nil || sels.IsEmpty() {
+		for i := offset; i < offset+length; i++ {
+			elem := vec.Get(i + bias)
+			if err = op(elem, i); err != nil {
+				break
+			}
+		}
+	} else {
+
+		selsArray := sels.ToArray()
+		end := offset + length
+		for _, rowId := range selsArray {
+			if int(rowId) < offset {
+				continue
+			} else if int(rowId) >= end {
+				break
+			}
+			elem := vec.Get(int(rowId) + bias)
+			if err = op(elem, int(rowId)); err != nil {
+				break
+			}
+		}
+	}
+	return
+}
+
+func (vec *CnTaeVector[T]) PPString(num int) string {
+	var w bytes.Buffer
+	_, _ = w.WriteString(fmt.Sprintf("[T=%s][Len=%d][Data=(", vec.GetType().String(), vec.Length()))
+	limit := vec.Length()
+	if num > 0 && num < limit {
+		limit = num
+	}
+	size := vec.Length()
+	long := false
+	if size > limit {
+		long = true
+		size = limit
+	}
+	for i := 0; i < size; i++ {
+		if vec.IsNull(i) {
+			_, _ = w.WriteString("null")
+			continue
+		}
+		if vec.GetType().IsVarlen() {
+			_, _ = w.WriteString(fmt.Sprintf("%s, ", vec.Get(i).([]byte)))
+		} else {
+			_, _ = w.WriteString(fmt.Sprintf("%v, ", vec.Get(i)))
+		}
+	}
+	if long {
+		_, _ = w.WriteString("...")
+	}
+	_, _ = w.WriteString(")]")
+	return w.String()
+}
+
+// TODO: --- I am not sure, if the below functions will work as expected
+
+func (vec *CnTaeVector[T]) Allocated() int {
```

Comment:
this does not make much sense.   Remove if we can, ...

Reply:
Sure. I will also check with @XuPeng-SH  to see its significance.

Right now it is used in [BlockReader](https://github.com/arjunsk/matrixone/blob/e5e451c6bee40b5c116bdd5e699dadb29d07059d/pkg/vm/engine/tae/dataio/blockio/read.go#L65).

I also don't know if this implementation is correct. I need someone to validate this.